### PR TITLE
Run mypy on the test cases and test scripts

### PR DESCRIPTION
**Pull request by AlexWaygood**
This PR extends `mypy_test.py` so that:

- On 3.6+, `mypy --strict` is run on the test cases in the `test_cases` directory
- On 3.9+, `mypy --strict` is run on the scripts in the `tests` directory. (`pytype_test.py` is excluded on Windows, since the test itself can't run on Windows and `pytype` can't be installed on Windows.)

### CODE REVIEW

File: tests/mypy_test.py [hunk 1]
```diff
@@ -280,6 +294,94 @@ def is_probably_stubs_folder(distribution: str, distribution_path: Path) -> bool
     return distribution != ".mypy_cache" and distribution_path.is_dir()
 
 
+class TestResults(NamedTuple):
+    exit_code: int
+    files_checked: int
+
+
+def test_typeshed(code: int, major: int, minor: int, args: argparse.Namespace) -> TestResults:
```
Comment (JelleZijlstra) [inline tests/mypy_test.py:302]: This function is rather long, what about splitting it into separate functions for the stdlib, third_party, tests/, and test_cases/?

Comment (AlexWaygood) [reply tests/mypy_test.py:302]: Sure!

File: tests/mypy_test.py [hunk 2]
```diff
@@ -280,6 +294,118 @@ def is_probably_stubs_folder(distribution: str, distribution_path: Path) -> bool
...
 
+class TestResults(NamedTuple):
+    exit_code: int
+    files_checked: int
+
+
+def test_stdlib(code: int, major: int, minor: int, args: argparse.Namespace) -> TestResults:
+    seen = {"__builtin__", "builtins", "typing"}  # Always ignore these.
+
+    # Test standard library files.
```
Comment (AlexWaygood) [inline tests/mypy_test.py:305]:
```suggestion
```

File: tests/mypy_test.py [hunk 3]
```diff
@@ -280,6 +294,117 @@ def is_probably_stubs_folder(distribution: str, distribution_path: Path) -> bool
...
+    return TestResults(code, files_checked)
+
+
+def test_the_test_scripts(code: int, major: int, minor: int, args: argparse.Namespace) -> TestResults:
+    files_to_test = list(Path("tests").rglob("*.py"))
+    if sys.platform == "win32":
+        files_to_test.remove(Path("tests/pytype_test.py"))
+    num_test_files_to_test = len(files_to_test)
+    flags = get_mypy_flags(args, major, minor, None, strict=True, test_suite_run=True)
+    print()
```
Comment (AlexWaygood) [inline tests/mypy_test.py:359]:
```suggestion
```

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**